README.md
# FS-EventBridge-JS [![Build Status](https://travis-ci.org/TomFrost/fs-eventbridge-js.svg?branch=master)](https://travis-ci.org/TomFrost/fs-eventbridge-js)
An ultra-lightweight Node.js client for the [FS-EventBridge](https://github.com/TomFrost/fs_eventbridge) filesystem change notification forwarder.

## Installation

```
npm install --save fs-eventbridge-js
```

## Usage

```javascript
const FSEventBridgeClient = require('fs-eventbridge-js')

const bridge = new FSEventBridgeClient({host: '192.168.1.12'})
bridge.start().then(() => {
  // process.cwd() is currently being monitored for changes.
  // These filesystem events will be replicated in the target
  // environment, regardless of whether it uses fsevents, inotify, or
  // something else.
})
```

### API

#### `new FSEventBridgeClient(opts)`
Creates a new client. You'll need one of these per server you want to connect to, per top-level folder to be watched.

Options:

- **host** _(Default: localhost)_: The hostname or IP address with which to connect
- **port** _(Default: 65056)_: The port with which to connect
- **watch** _(Default: `process.cwd()`)_: The directory to be watched for changes
- **recursive** _(Default: true)_: If true, all subdirectories and their files will also be watched
- **persistent** _(Default: false)_: If true, prevents the Node.js process from ending while the client is monitoring for file changes.
- **ignoreMs** _(Default: 250)_: The number of milliseconds during which any further change notifications for a given file are ignored. This helps reduce load for rapidly-changing files, as well as prevents an infinite loop of reporting changes that were due to the last reported change.
- **ignoreHidden** _(Default: true)_: Ignores file changes within hidden files and folders if true

#### `client.start()`
Connects to the FS-EventBridge server and starts monitoring the filesystem for changes. Returns a Promise that resolves whenever a connection has been established, rejecting on connection failure. A resolved promise is returned if this is called during an active connection.

#### `client.stop()`
Stops watching for filesystem events and disconnects from the server. Has no effect if called when there is no active connection.

#### `client.connected`
A getter that provides `true` if there is an active connection.

### Events

#### `error_connection` Error
Emitted whenever the connection to the FS-EventBridge server has an unexpected issue. The server has been stopped when this event is fired, and `start()` can be called on the same instance if desired.

#### `error_command` Error
Emitted whenever the FS-EventBridge server communicates an error. These have no long-term impact on the client, but may signify that a file change event did not propagate to the remote operating system.

#### `error_file` Error
Emitted when the client cannot `stat` a changed file. These have no long-term impact on the client, but may signify that a file change event did not propagate to the remote operating system.

#### `response` string
Emitted when the server sends any response. This is useful for debug logging.

## License
FS-EventBridge-JS is distributed under the ISC license. See `LICENSE` file for details.

## Credits
FS-EventBridge-JS was originally created at [TechnologyAdvice](http://technologyadvice.com) in Nashville, TN.


CHANGELOG.md
# FS-EventBridge-JS Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

## [Development]
Nothing yet!

## [v0.2.0]
### Added
- `ignoreMs` option to prevent infinite loops in NFS-share use cases, and reduce load of rapidly-changing files
- `ignoreHidden` option to ignore changes in hidden files and folders

### Changed
- `client.start()` will now properly reject if the connection fails.

## v0.1.0
### Added
- Initial release

[Development]: https://github.com/TomFrost/Squiss/compare/v0.2.0...HEAD
[v0.2.0]: https://github.com/TomFrost/Squiss/compare/v0.1.0...v0.2.0


test/setup.js
/*
 * Copyright (c) 2017 Tom Shawver
 */

'use strict'

// Unit/Integration tests
const path = require('path')
const chai = require('chai')
const mod = require('module')
global.should = chai.should()

// importing files with ../../../../../.. makes my brain hurt
process.env.NODE_PATH = path.join(__dirname, '..') + path.delimiter + (process.env.NODE_PATH || '')
mod._initPaths()


test/src/index.spec.js
/*
 * Copyright (c) 2017 Tom Shawver
 */

/* global should, describe, it, afterEach */

'use strict'

const FSEventBridgeClient = require('src/index')
const net = require('net')
const fs = require('fs')
const path = require('path')

let inst
let server

describe('FSEventBridgeClient', () => {
  afterEach(() => {
    if (inst) inst.stop()
    if (server) server.close()
  })
  it('instantiates without error', () => {
    new FSEventBridgeClient()
  })
  it('connects to a server', (done) => {
    server = net.createServer(() => done())
    server.listen(65056)
    inst = new FSEventBridgeClient()
    inst.start()
  })
  it('watches for file changes and sends change commands', (done) => {
    const nowMs = Date.now()
    const file = path.join(process.cwd(), 'package.json')
    server = net.createServer(conn => {
      conn.on('data', data => {
        data.toString().should.equal(`CHANGE ${file}\n`)
        done()
      })
    })
    server.listen(65056)
    inst = new FSEventBridgeClient()
    inst.start().then(() => {
      inst.connected.should.equal(true)
      fs.utimes(file, nowMs, nowMs, err => {
        if (err) done(err)
      })
    }).catch(done)
  })
})


src/index.js
/*
 * Copyright (c) 2017 Tom Shawver
 */

'use strict'

const fs = require('fs')
const net = require('net')
const path = require('path')
const EventEmitter = require('events').EventEmitter

const DEFAULT_PORT = 65056
const errRegex = /^ERR\s(.+)$/

class FSEventBridgeClient extends EventEmitter {

  /**
   * Constructs a new FSEventBridgeClient
   * @param {Object} opts An option mapping
   * @param {string} [opts.host="localhost"] The hostname or IP address with which to connect
   * @param {number} [opts.port=65056] The port with which to connect
   * @param {string} [opts.watch=process.cwd()] The directory to be watched for changes
   * @param {boolean} [opts.recursive=true] If true, all subdirectories and their files will also be watched
   * @param {boolean} [opts.persistent=false] If true, If true, prevents the Node.js process from ending while
   * the client is monitoring for file changes.
   * @param {number} [opts.ignoreMs=250] The number of milliseconds during which any further change notifications
   * for a given file are ignored. This helps reduce load for rapidly-changing files, as well as prevents an infinite
   * loop of reporting changes that were due to the last reported change.
   * @param {boolean} [opts.ignoreHidden=true] Ignores file changes within hidden files and folders if true
   */
  constructor(opts) {
    super()
    const defaults = {
      host: 'localhost',
      port: DEFAULT_PORT,
      watch: process.cwd(),
      recursive: true,
      persistent: false,
      ignoreMs: 250,
      ignoreHidden: true
    }
    this._conn = {end: () => {}}
    this._opts = Object.assign({}, defaults, opts || {})
    this._connected = false
    this._ignore = {}
  }

  get connected() {
    return this._connected
  }

  /**
   * Connects to the FS-EventBridge server and starts listening for and forwarding filesystem events.
   * @returns {Promise} Resolves upon successful connection
   */
  start() {
    if (this._connected) return Promise.resolve()
    return this._connect().then(() => {
      this._connected = true
      const watchOpts = {
        persistent: this._opts.persistent,
        recursive: this._opts.recursive
      }
      fs.watch(this._opts.watch, watchOpts, (type, file) => {
        if (this._ignore[file] || FSEventBridgeClient.isHidden(file)) return undefined
        this._ignore[file] = true
        setTimeout(() => { delete this._ignore[file] }, this._opts.ignoreMs)
        this.emit('changed', file)
        const abs = path.join(process.cwd(), file)
        this._conn.write(`CHANGE ${abs}\n`)
      })
    })
  }

  /**
   * Ends the connection to the FS-EventBridge server and stops listening for filesystem events.
   */
  stop() {
    if (this._connected) {
      this._conn.end()
      fs.unwatchFile(this._opts.watch)
      this._connected = false
    }
  }

  _handleLine(line) {
    this.emit('response', line)
    const errMatch = line.match(errRegex)
    if (errMatch) this.emit('command_error', new Error(errMatch[1]))
  }

  _connect() {
    let strBuf = ''
    return new Promise((resolve, reject) => {
      this._conn = net.connect(this._opts.port, this._opts.host, resolve)
      this._conn.once('end', () => this.stop())
      this._conn.once('error', (e) => {
        this.stop()
        this.emit('connection_error', e)
        reject(e)
      })
      this._conn.on('data', (buf) => {
        strBuf += buf.toString()
        const lines = strBuf.split('\n')
        strBuf = lines.pop()
        lines.forEach(line => this._handleLine(line))
      })
    })
  }

  static isHidden(file) {
    return file[0] === '.' || file.includes(`${path.sep}.`)
  }
}

module.exports = FSEventBridgeClient
